### ISSUE TITLE
{RFC} Requirement of Actions to enforce Code Formatting.

### ISSUE THREAD

**Issue by AbhijithGanesh** [CiviWiki/OpenCiviWiki#1033]
### Idea summary

We need a code-formatter

### Further details

I think it is high time we start using flake, autopep8 or black. I am personally inclined to use black but if there are standards which the project adapts, I shall make amends.

**Comment by gorkemarslan** [comment]
`flake8` has already been included in `requirements/dev.txt`. Before committing, we are all supposed to check the format with the command of `flake8` on the console. 

I remember there was a bot that checks code format for PRs in the project. I don't know we still have it. Do you know where I can find bots present in the project?

**Comment by AbhijithGanesh** [comment]
I think GitHub Actions take care of code formatting, we can configure yml files for workflows and get it running,

**Comment by gorkemarslan** [comment]
That is a good idea. :) Could you edit your issue to specify this feature?

**Comment by AbhijithGanesh** [comment]
Edited

**Comment by AbhijithGanesh** [comment]
I would like to resolve this immediately, it is important and it'd be nice if I get it done.

**Comment by AbhijithGanesh** [comment]
Worked on this and created a PR, please review @Brylie and @gorkemarslan

**Comment by brylie** [comment]
> I remember there was a bot that checks code format for PRs in the project. I don't know we still have it.

@pep8speaks bot scans all pull requests for lint issues.

> I am personally inclined to use black

I agree we should add `black` as a development dependency.

> I think GitHub Actions take care of code formatting, we can configure yml files for workflows and get it running,

What are the security implications of having untrusted code with write permissions on this repository? How will we be able to audit the code changes made by GitHub Actions?

**Comment by AbhijithGanesh** [comment]
I am unaware of methods of auditing with GitHub Actions but there is definitely a provision for that.

As for the security concerns, The GitHub Actions _will not be having write access_. It will not merge code without the approval of the staff. The reason why it should be implemented is that it can do more than pep8 scans. It can check on JS and even other types of work such as documentation if configured well. Its upto us to decide though.

**Comment by brylie** [comment]
OK, sounds good. So it would add a notice in the pull request? Which specific GitHub action are we considering? I'll read the documentation for the relevant action.

**Comment by brylie** [comment]
For example, are we considering adding [rickstaa/action-black](https://github.com/rickstaa/action-black/)?

**Comment by AbhijithGanesh** [comment]
We can. There are countless examples.I was of the opinion that we'll be using flake as Gorkem Arslan mentioned. If we need another formatter we can get it.

**Comment by brylie** [comment]
@AbhijithGanesh please suggest a specific GitHub action so that we can audit the code/documentation.

**Comment by brylie** [comment]
> I was of the opinion that we'll be using flake

If that is the case, then no changes are needed since @pep8speaks is already auditing our pull requests

**Comment by AbhijithGanesh** [comment]
Yeah but GitHub Actions act as Pull Request checks,it shows on the PR tab and it is more convenient to use too, It is a part of standard practices as well.

**Comment by brylie** [comment]
OK, that makes sense.

We still need to identify and discuss one or more potential GitHub actions related to this idea. Would you mind suggesting which GitHub action(s) might resolve this feature request?

**Comment by wassafshahzad** [comment]
@brylie can I have this?

**Comment by brylie** [comment]
@wassafshahzad, assigned! Thanks :smiley:

**Comment by AbhijithGanesh** [comment]
Can I review this once Wassafshahzad is done?

**Comment by AbhijithGanesh** [comment]
I would like to see the PR and know what kind of checks will be implemented.

**Comment by brylie** [comment]
> Can I review this once Wassafshahzad is done?

Sure. @wassafshahzad, would you mention this issue ("Closes #1033") and @AbhijithGanesh in your pull request?

**Comment by wassafshahzad** [comment]
Sure

**Comment by wassafshahzad** [comment]
Created a new PR @brylie  kindly assign @AbhijithGanesh  to i and sorry I couldn't be active I was held up in my job.

**Comment by AbhijithGanesh** [comment]
Thanks for informing and your wonderful efforts @wassafshahzad

**Comment by brylie** [comment]
@wassafshahzad the PR looks excellent! As mentioned above, we are using `black` as a development dependency for this project, which is compatible with `flake8`. So please make the small suggested changes to the `flake8` configuration, so it is harmonious with the `black` defaults.

**Comment by AbhijithGanesh** [comment]
Agreed 👍🏻
